geomstats/geometry/waldspace.py

```diff
@@ -0,0 +1,437 @@
+                old_splits = new_splits
+            return old_splits
+
+        def _check_separability(splits, labels):
+            """Checks if exists at least one split between each pair of labels."""
+            return gs.all([gs.any([sp.separates(_u, _v) for sp in splits]) for _u, _v in
+                           it.combinations(labels, 2)])
+
+        def _delete_random_edges(splits, labels, probability):
+            """Given set of ``splits`` of set of ``labels``, delete random splits."""
+            if probability == 1:
+                return splits
+            # random decide whether to delete splits, starting from the last split
+            print(f"Test 1 -- splits are {splits}.")
+            for i in reversed(range(len(splits))):
+                # in this case, delete split, if allowed
+                if gs.random.rand(1) > probability:
+                    splits_copy = splits.copy()
+                    splits_copy.pop(i)
+                    # if allowed (i.e. exists split separating every pair of labels),
+                    # delete the split, else not (don't change splits)
+                    if _check_separability(splits=splits_copy, labels=labels):
+                        splits = splits_copy
+            print(f"Test 2 -- splits are {splits}.")
+            return splits
+
+        def _generate_wald(_prob):
+            """Generates a random wald."""
+            # generate random partition
+            partition = _generate_partition(_prob=_prob)
+            # generate random sets of splits for each component of the partition
+            split_sets = [_generate_splits(labels=_part) for _part in partition]
+            # delete random splits of those sets of splits
+            split_sets = [
+                _delete_random_edges(splits=_splits, labels=_part, probability=_prob)
+                for _part, _splits in zip(partition, split_sets)]
+            # create the structure for the wald, that is partition + sets of splits
+            st = Structure(n=self.n, partition=partition, split_sets=split_sets)
+            # generate random weights for the edges
+            x = gs.random.uniform(size=(len(st.unravel(st.split_sets)),), low=0, high=1)
+            # TODO element wise minimum of arrays, minimum needed. gs.minimum not impl.
+            x = gs.minimum(gs.maximum(btol, x), 1 - btol)
+            # create the wald
+            return Wald(n=self.n, st=st, x=x)
+
+        # generate samples:
+        prob = prob ** (1 / (self.n - 1))
+        if n_samples == 1:
+            sample = _generate_wald(_prob=prob)
+            print(sample)
+            print(sample.x.shape)
+            return sample.corr
+        return gs.array([_generate_wald(_prob=prob).corr for _ in range(n_samples)])
+
+
+def equivalence_partition(group, relation):
+    """Partitions a set of objects into equivalence classes
+    Taken from
+    'https://stackoverflow.com/questions/38924421/is-there-a-standard-way-to-partition-a
```

Comment:
refer to it as reference or in the see also section at the bottom of the docstring